Enable SSE2 compression path to work on MSVC

TrianglesPCT opened the pull request:
The compile time detection uses the `__SSE2__ `predefined macro, but this does not exist on MSVC--resulting it that compiler always using scalar path.

On MSVC you can detect SSE support by checking for `_M_AMD64`, as when targeting x64 it requires SSE2 as a baseline, this patch adds this check.

TrianglesPCT committed:
Add files via upload

msvc suport
avx2 path

TrianglesPCT committed:
Add files via upload

msvc

TrianglesPCT committed:
Add files via upload

TrianglesPCT committed:
Update zstd_lazy.c

use 8bit

TrianglesPCT reviewed (commented):
.

TrianglesPCT committed:
Update zstd_lazy.c

TrianglesPCT committed:
Update zstd_lazy.c

switch to unaligned load as I don't know if buffer will always be aligned to 32 bytes, and compilers aside from MSVC might actually use aligned loads

TrianglesPCT reviewed (commented):
Switch to unaligned load

Cyan4973 commented:
Thanks @TrianglesPCT ,

did you attempt to measure performance differences provided by your PR ?

terrelln commented on lib/compress/zstd_lazy.c:
```diff
@@ -921,6 +921,28 @@ static ZSTD_VecMask ZSTD_Vec256_cmpMask8(ZSTD_Vec256 x, ZSTD_Vec256 y) {
   sndMask = ZSTD_Vec128_cmpMask8(x.snd, y.snd);
   return fstMask | (sndMask << 16);
 }
+#else//AVX2 
+typedef struct {
+	__m256i v;
+} ZSTD_Vec256;
+
+static ZSTD_Vec256 ZSTD_Vec256_read(const void* const ptr) {
+	ZSTD_Vec256 v;
+    v.v = _mm256_loadu_si256((const __m256i*)ptr);
+	return v;
+}
+
+static ZSTD_Vec256 ZSTD_Vec256_set8(BYTE val) {
+	ZSTD_Vec256 v;
+    v.v = _mm256_set1_epi8((char)val);
+	return v;
+}
+
+static ZSTD_VecMask ZSTD_Vec256_cmpMask8(ZSTD_Vec256 x, ZSTD_Vec256 y) {
+    return (ZSTD_VecMask)_mm256_movemask_epi8(_mm256_cmpeq_epi8(x.v, y.v));
+}
+
+#endif
```
Can you submit this portion as a separate PR please?

It should be good go go with some minor formatting changes, but this portion requires performance testing, where the MSVC fix is easy to accept.

I know I measured the performance of AVX2 vs SSE, and it was slightly better. But, I didn't see a huge speed-up.

TrianglesPCT replied:
Yea I'll remove this and submit it seperate

terrelln commented on lib/compress/zstd_lazy.c:
```diff
@@ -921,6 +921,28 @@ static ZSTD_VecMask ZSTD_Vec256_cmpMask8(ZSTD_Vec256 x, ZSTD_Vec256 y) {
+
+static ZSTD_Vec256 ZSTD_Vec256_set8(BYTE val) {
+	ZSTD_Vec256 v;
+    v.v = _mm256_set1_epi8((char)val);
```
nit: formatting

terrelln commented on lib/compress/zstd_lazy.c:
```diff
@@ -921,6 +921,28 @@ static ZSTD_VecMask ZSTD_Vec256_cmpMask8(ZSTD_Vec256 x, ZSTD_Vec256 y) {
   sndMask = ZSTD_Vec128_cmpMask8(x.snd, y.snd);
   return fstMask | (sndMask << 16);
 }
+#else//AVX2 
```
nit: We are C90, so you need to do `/* AVX2 */`

TrianglesPCT committed:
Update zstd_lazy.c

Switch to other comment style

TrianglesPCT committed:
Update zstd_lazy.c

Remove the AVX2 part

TrianglesPCT committed:
Update zstd_lazy.c

add space

TrianglesPCT committed:
Add files via upload

AVX2

TrianglesPCT committed:
Update zstd_lazy.c

It put the changes back when I tried to make a separate pull request, i don't understand githubs interface at all.

TrianglesPCT commented:
> 
> 
> Thanks @TrianglesPCT ,
> 
> did you attempt to measure performance differences provided by your PR ?

I removed the AVX2 stuff for a later PR

Cyan4973 commented:
Well, even without `AVX2`,
this PR is presumed to help performance, because it's supposed to help `SSE2` detection ?

TrianglesPCT commented:
All it is doing is normalizing MSVC to target the same code path as GCC/Clang when SSE2 is set as a requirement for your compile target. Not adding any new code..

 The `defined(__SSE2__)` check doesn't work on MSVC because that define doesn't exist even if SSE2 is enabled.

 If you want specific performance comparison I can do that might be few days until I have time for that.

terrelln commented:
> If you want specific performance comparison I can do that might be few days until I have time for that.

That would be great! But that doesn't need to block merging this PR. You can easily measure performance with the builtin benchmark tool.

```
zstd -b5e12 silesia.tar
```

ghost commented:
Will there be a hot fix for v1.5.0?

wolfpld commented:
> You can easily measure performance with the builtin benchmark tool.

The difference is rather spectacular. Results for 5950X.

Before:
```
 5#silesia.tar       : 211957760 ->  63806593 (3.322), 111.0 MB/s , 977.5 MB/s
 6#silesia.tar       : 211957760 ->  62980544 (3.365), 107.9 MB/s ,1001.6 MB/s
 7#silesia.tar       : 211957760 ->  61482185 (3.447),  73.1 MB/s ,1055.9 MB/s
 8#silesia.tar       : 211957760 ->  60914308 (3.480),  57.2 MB/s ,1076.1 MB/s
 9#silesia.tar       : 211957760 ->  59928587 (3.537),  52.1 MB/s ,1100.2 MB/s
10#silesia.tar       : 211957760 ->  59296205 (3.575),  50.3 MB/s ,1103.6 MB/s
11#silesia.tar       : 211957760 ->  59152928 (3.583),  47.7 MB/s ,1106.8 MB/s
12#silesia.tar       : 211957760 ->  58640204 (3.615),  30.6 MB/s ,1123.8 MB/s
```

After:
```
 5#silesia.tar       : 211957760 ->  63806593 (3.322), 160.8 MB/s , 980.0 MB/s
 6#silesia.tar       : 211957760 ->  62980544 (3.365), 153.3 MB/s ,1004.7 MB/s
 7#silesia.tar       : 211957760 ->  61482185 (3.447), 111.9 MB/s ,1056.3 MB/s
 8#silesia.tar       : 211957760 ->  60914308 (3.480),  90.9 MB/s ,1079.5 MB/s
 9#silesia.tar       : 211957760 ->  59928587 (3.537),  74.7 MB/s ,1101.9 MB/s
10#silesia.tar       : 211957760 ->  59296205 (3.575),  71.3 MB/s ,1107.0 MB/s
11#silesia.tar       : 211957760 ->  59152928 (3.583),  65.0 MB/s ,1113.2 MB/s
12#silesia.tar       : 211957760 ->  58640204 (3.615),  51.3 MB/s ,1129.6 MB/s
```

senhuang42 reviewed (approved):
This is a great fix, thanks for the contribution!
